src/expr/src/explain/mod.rs

```diff
@@ -46,6 +54,88 @@ pub struct ExplainSinglePlan<'a, T> {
     pub plan: AnnotatedPlan<'a, T>,
 }
 
+/// Carries metadata about the possibility of MFP pushdown for a source.
+/// (Likely to change, and only emitted when a context flag is enabled.)
+#[allow(missing_debug_implementations)]
+pub struct PushdownInfo {
+    /// Pushdown-able columns in the source.
+    pub cols: Vec<usize>,
+}
+
+impl<C: AsMut<Indent>> DisplayText<C> for PushdownInfo {
+    fn fmt_text(&self, f: &mut Formatter<'_>, ctx: &mut C) -> std::fmt::Result {
+        if !self.cols.is_empty() {
+            writeln!(
+                f,
+                "{}pushdown=(#{})",
+                ctx.as_mut(),
+                self.cols.iter().join(", #")
+            )?;
+        }
+        Ok(())
+    }
+}
+
+#[allow(missing_debug_implementations)]
+pub struct ExplainMultiPlanSource<'a> {
+    pub id: String,
+    pub op: &'a MapFilterProject,
+    pub pushdown_info: Option<PushdownInfo>,
+}
+
+impl<'a> ExplainMultiPlanSource<'a> {
+    pub fn new(
+        id: String,
+        op: &'a MapFilterProject,
+        context: &ExplainContext<'a>,
+    ) -> ExplainMultiPlanSource<'a> {
+        let pushdown_info = if context.config.mfp_pushdown {
```

Comment:
I think in the long term we would want to change the meaning of this option, so when `mfp_pushdown` is not `true` we don't render the `sources` blocks at all. The default can be `true`.